soc/nordic/nrf54h/gpd/gpd.c

```diff
@@ -0,0 +1,273 @@
+		struct gpd_onoff_manager *gpd_mgr = context;
+
+		switch (p_evt->type) {
+		case NRFS_GDPWR_REQ_APPLIED:
+			gpd_mgr->notify(&gpd_mgr->mgr, 0);
+			break;
+		default:
+			LOG_ERR("nRFs GDPWR request not applied");
+			gpd_mgr->notify(&gpd_mgr->mgr, -EIO);
+			break;
+		}
+	} else {
+		switch (p_evt->type) {
+		case NRFS_GDPWR_REQ_APPLIED:
+			atomic_set_bit(&gpd_service_status, GPD_SERVICE_REQ_OK);
+			break;
+		default:
+			LOG_ERR("nRFs GDPWR request not applied");
+			atomic_set_bit(&gpd_service_status, GPD_SERVICE_REQ_ERR);
+			break;
+		}
+	}
+}
+
+static void start(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	struct gpd_onoff_manager *gpd_mgr = CONTAINER_OF(mgr, struct gpd_onoff_manager, mgr);
+
+	gpd_mgr->notify = notify;
+
+	if (!atomic_test_bit(&gpd_service_status, GPD_SERVICE_READY)) {
+		notify(mgr, 0);
+	} else {
+		nrfs_err_t err;
+
+		err = nrfs_gdpwr_power_request(gpd_mgr->id, GDPWR_POWER_REQUEST_SET, gpd_mgr);
+		if (err != NRFS_SUCCESS) {
+			LOG_ERR("nRFs GDPWR request failed (%d)", err);
+			notify(mgr, -EIO);
+		}
+	}
+}
+
+static void stop(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	struct gpd_onoff_manager *gpd_mgr = CONTAINER_OF(mgr, struct gpd_onoff_manager, mgr);
+
+	gpd_mgr->notify = notify;
+
+	if (!atomic_test_bit(&gpd_service_status, GPD_SERVICE_READY)) {
+		notify(mgr, 0);
+	} else {
+		nrfs_err_t err;
+
+		err = nrfs_gdpwr_power_request(gpd_mgr->id, GDPWR_POWER_REQUEST_CLEAR, gpd_mgr);
+		if (err != NRFS_SUCCESS) {
+			LOG_ERR("nRFs GDPWR request failed (%d)", err);
+			notify(mgr, -EIO);
+		}
+	}
+}
+
+int nrf_gpd_request(uint8_t id)
+{
+	int ret;
+	struct onoff_client client;
+	struct gpd_onoff_manager *gpd_mgr;
+
+	gpd_mgr = get_mgr(id);
+	if (gpd_mgr == NULL) {
+		return -EINVAL;
+	}
+
+	if (atomic_test_bit(&gpd_service_status, GPD_SERVICE_ERROR)) {
+		LOG_ERR("GPD service did not initialize properly");
+		return -EIO;
+	}
+
+	sys_notify_init_spinwait(&client.notify);
+
+	onoff_request(&gpd_mgr->mgr, &client);
+
+	while (sys_notify_fetch_result(&client.notify, &ret) == -EAGAIN) {
+	}
+
+	return ret;
+}
+
+int nrf_gpd_release(uint8_t id)
+{
+	struct gpd_onoff_manager *gpd_mgr;
+
+	gpd_mgr = get_mgr(id);
+	if (gpd_mgr == NULL) {
+		return -EINVAL;
+	}
+
+	if (atomic_test_bit(&gpd_service_status, GPD_SERVICE_ERROR)) {
+		LOG_ERR("GPD service did not initialize properly");
+		return -EIO;
+	}
+
+	return onoff_release(&gpd_mgr->mgr);
+}
+
+static int nrf_gpd_pre_init(void)
+{
+	int ret;
+
+	ret = onoff_manager_init(&fast_active1.mgr, &transitions);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = onoff_manager_init(&slow_active.mgr, &transitions);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = onoff_manager_init(&slow_main.mgr, &transitions);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int nrf_gpd_post_init(void)
+{
+	nrfs_err_t err;
+	int ret;
+
+	ret = nrfs_backend_wait_for_connection(K_NO_WAIT);
+	if (ret < 0) {
+		return ret;
```

Comment:
I think that `atomic_set_bit(&gpd_service_status, GPD_SERVICE_ERROR);` is missing here.